File: src/services/lsp.ts
import { ipcInvoke } from "./ipc";
import { getLocusRuntime, type RuntimeUnsubscribe } from "./locusRuntime";

export type LspKind = "omnisharp";

interface LspMessageEvent {
  session_id: string;
  message: unknown;
}

interface LspExitEvent {
  session_id: string;
  code: number | null;
  error: string | null;
}

export interface LspSession {
  readonly id: string;
  send(message: unknown): Promise<void>;
  onMessage(handler: (message: unknown) => void): Promise<RuntimeUnsubscribe>;
  onExit(handler: (info: { code: number | null; error: string | null }) => void): Promise<RuntimeUnsubscribe>;
  stop(): Promise<void>;
}

export async function startLsp(kind: LspKind, workspaceDir: string): Promise<LspSession> {
  const id = await ipcInvoke<string>("lsp_start", { kind, workspaceDir });
  const runtime = getLocusRuntime();

  return {
    id,
    send(message) {
      return ipcInvoke<void>("lsp_send", { sessionId: id, message });
    },
    onMessage(handler) {
      return runtime.subscribe<LspMessageEvent>(`lsp:${id}:message`, (payload) => {
        handler(payload.message);
      });
    },
    onExit(handler) {
      return runtime.subscribe<LspExitEvent>(`lsp:${id}:exit`, (payload) => {
        handler({ code: payload.code, error: payload.error });
      });
    },
    stop() {
      return ipcInvoke<void>("lsp_stop", { sessionId: id });
    },
  };
}

File: src-tauri/src/commands/lsp.rs
use std::sync::Arc;

use tauri::{AppHandle, State};

use crate::error::AppError;
use crate::lsp::{LspKind, LspManager};

#[tauri::command]
pub async fn lsp_start(
    kind: LspKind,
    workspace_dir: String,
    app: AppHandle,
    manager: State<'_, Arc<LspManager>>,
) -> Result<String, AppError> {
    manager.start(&app, kind, workspace_dir).await
}

#[tauri::command]
pub async fn lsp_send(
    session_id: String,
    message: serde_json::Value,
    manager: State<'_, Arc<LspManager>>,
) -> Result<(), AppError> {
    manager.send(&session_id, message).await
}

#[tauri::command]
pub async fn lsp_stop(
    session_id: String,
    manager: State<'_, Arc<LspManager>>,
) -> Result<(), AppError> {
    manager.stop(&session_id).await
}
